### ISSUE TITLE
Questions regarding lactylation modifications in DIANN version 1.9.1 windows

### ISSUE THREAD

**Issue by ZHBHSMILE** [vdemichev/DiaNN#1248]
Hi Vadim,

Thank you, as always, for developing such an excellent tool.

I am currently analyzing lactylation and have encountered an issue in DIANN version 1.9.1:

   **My Questions**
   1.   Lactylation modifications are frequently detected only once at the end of Modified.Sequence,  as shown in the attached image. Our goal is to exclude sequences where lactylation appears solely at the sequence’s end，because of the position after trypsin cleavage, lactylation modifications should not occur. In the advanced options window of the GUI,  is there a command to exclude sequences where lactylation appears only at the end?
  ![qustion](https://github.com/user-attachments/assets/efef9398-95c8-4a6c-b7f8-5158690d9680)

   2. Expected Count for Lactylation: What is the typical number of Modified.Sequence entries expected to contain lactylation? With 172,893 unique sequences, I am concerned this may be an error. Could you confirm if this quantity is within an expected range?
   The total unique Modified.Sequence count I obtained is 172,893, which seems unusually high. I suspect there may be an error in my commands:

  ```
  data <- read.delim("2024_11_4_SP_Homo_sapiens_report.pr_matrix.tsv", header = TRUE, check.names = FALSE)
  s <- data$Modified.Sequence %>% unique()         # Total unique sequences: 172,893
  s1 <- data$Modified.Sequence[str_detect(data$Modified.Sequence, "2114")]  # Sequences with lactylation: 153,072
  s2 <- s1[(!str_detect(s1, ".*2114.*2114.*")) & str_detect(s1, ".*2114\\)$")] %>% length()  # Lactylation only at sequence end: 34,482
  ```
command in log.tsv:
  ```
 diann.exe --f E:\project\12F236DP\F236DP\F236DP-N1.raw  --f E:\project\12F236DP\F236DP\F236DP-N2.raw  --f E:\project\12F236DP\F236DP\F236DP-N3.raw  --f E:\project\12F236DP\F236DP\F236DP-N4.raw  --f E:\project\12F236DP\F236DP\F236DP-N5.raw  --f E:\project\12F236DP\F236DP\F236DP-N6.raw  --f E:\project\12F236DP\F236DP\F236DP-N7.raw  --f E:\project\12F236DP\F236DP\F236DP-N8.raw  --f E:\project\12F236DP\F236DP\F236DP-N9.raw  --f E:\project\12F236DP\F236DP\F236DP-N10.raw  --f E:\project\12F236DP\F236DP\F236DP-N11.raw  --f E:\project\12F236DP\F236DP\F236DP-N12.raw  --f E:\project\12F236DP\F236DP\F236DP-N13.raw  --f E:\project\12F236DP\F236DP\F236DP-N14.raw  --f E:\project\12F236DP\F236DP\F236DP-N15.raw  --f E:\project\12F236DP\F236DP\F236DP-T1.raw  --f E:\project\12F236DP\F236DP\F236DP-T2.raw  --f E:\project\12F236DP\F236DP\F236DP-T3.raw  --f E:\project\12F236DP\F236DP\F236DP-T4.raw  --f E:\project\12F236DP\F236DP\F236DP-T5.raw  --f E:\project\12F236DP\F236DP\F236DP-T6.raw  --f E:\project\12F236DP\F236DP\F236DP-T7.raw  --f E:\project\12F236DP\F236DP\F236DP-T8.raw  --f E:\project\12F236DP\F236DP\F236DP-T9.raw  --f E:\project\12F236DP\F236DP\F236DP-T10.raw  --f E:\project\12F236DP\F236DP\F236DP-T11.raw  --f E:\project\12F236DP\F236DP\F236DP-T12.raw  --f E:\project\12F236DP\F236DP\F236DP-T13.raw  --f E:\project\12F236DP\F236DP\F236DP-T14.raw  --f E:\project\12F236DP\F236DP\F236DP-T15.raw  --lib  --threads 30 --verbose 1 --out E:\project\12F236DP\F236DP_search2\2024_11_4_SP_Homo_sapiens_report.tsv --qvalue 0.01 --matrices --out-lib E:\project\12F236DP\F236DP_search2\SP_Homo_sapiens_report-lib.parquet --gen-spec-lib --predictor --fasta E:\project\12F236DP\F236DP\20220122SP_Homo_sapiens.fasta --fasta-search --min-fr-mz 200 --max-fr-mz 1800 --met-excision --min-pep-len 7 --max-pep-len 30 --min-pr-mz 300 --max-pr-mz 1800 --min-pr-charge 1 --max-pr-charge 4 --cut K*,R* --missed-cleavages 2 --unimod4 --var-mods 1 --var-mod UniMod:35,15.994915,M --var-mod UniMod:1,42.010565,*n --use-quant --peptidoforms --reanalyse --relaxed-prot-inf --rt-profiling --var-mod UniMod:2114,72.021129,K  --var-mods 4 
  Could you provide guidance on how to configure these options to achieve the correct results?
  ```

   3. Manual Deletion of Modified.Sequence Entries: If the GUI does not support this exclusion, would manually deleting Modified.Sequence entries with lactylation at the end help ensure accuracy in the pr.matrix.tsv file?

   4. Impact on pg.matrix.tsv: If we manually remove these sequences from pr.matrix.tsv, will this affect the pg.matrix.tsv file? Is it necessary to preserve the raw pg.matrix.tsv without any modifications?

Thank you very much for your guidance.

Best wishes,
zplv

**Comment by vdemichev** [comment]
Hi zplv,

> In the advanced options window of the GUI, is there a command to exclude sequences where lactylation appears only at the end?

No, the way to do this would be to edit the spectral library in R or Python to exclude any precursors.

> What is the typical number of Modified.Sequence entries expected to contain lactylation?

Don't know, never worked with this PTM :)

> The total unique Modified.Sequence count I obtained is 172,893, which seems unusually high.

Please share the DIA-NN logs, including for the library generation.

In general, please base your analysis exclusively on the main DIA-NN report in .parquet format, matrices are there just for quick analyses in Excel.

Best,
Vadim

**Comment by vdemichev** [comment]
Hi zplv,

Please address both warnings in the log, it is essential.

Best,
Vadim

**Comment by ZHBHSMILE** [comment]
Hi Vadim,

   Thank you for pointing that out. I’ll address both warnings in the log as requested.


Best wishes,
zplv

**Comment by xiaodfeng** [comment]
I came across similar questions. With DIA-NN, I only detected three lactylation modifications peptides. Is this a reasonable number? WIth DDA+Maxquant using Maxquant, we are able to detect 70. See attached log file for reference.
`
DIA-NN 2.1.0 Academia  (Data-Independent Acquisition by Neural Networks)
Compiled on Mar 23 2025 15:49:03
Current date and time: Tue Jun 24 17:18:50 2025
Logical CPU cores: 64
DIA-NN-2.1.0-Academia-Linux/diann-2.1.0/diann-linux --cut K*,R*,!*P --dir OptDiann/mzML/ --fasta 2023-02-17-decoys-contam-uniprot_human.fasta.fas --fasta-search --fixed-mod UniMod:4,57.021464,C --gen-spec-lib --lib  --matrices --missed-cleavages 1 --met-excision --min-fr-mz 200 --max-fr-mz 1800  --min-pep-len 7 --max-pep-len 50 --min-pr-mz 300 --max-pr-mz 1800 --min-pr-charge 1 --max-pr-charge 4 --no-quant-files --predictor --qvalue 0.01 --relaxed-prot-inf --reanalyse --smart-profiling --threads 16 --var-mods 5 --var-mod UniMod:1,42.010565,*n --var-mod UniMod:35,15.994915,M --var-mod UniMod:2114,72.021129,K --verbose 1 

In silico digest will involve cuts at K*,R*
But excluding cuts at *P
DIA-NN will carry out FASTA digest for in silico lib generation
Modification UniMod:4 with mass delta 57.0215 at C will be considered as fixed
A spectral library will be generated
Precursor/protein x samples expression level matrices will be saved along with the main report
Maximum number of missed cleavages set to 1
N-terminal methionine excision enabled
Min fragment m/z set to 200
Max fragment m/z set to 1800
Min peptide length set to 7
Max peptide length set to 50
Min precursor m/z set to 300
Max precursor m/z set to 1800
Min precursor charge set to 1
Max precursor charge set to 4
.quant files will not be saved to the disk
Deep learning will be used to generate a new in silico spectral library from peptides provided
Output will be filtered at 0.01 FDR
Heuristic protein grouping will be used, to reduce the number of protein groups obtained; this mode is recommended for benchmarking protein ID numbers, GO/pathway and system-scale analyses
MBR enabled; .quant files will only be saved to disk during the first pass
When generating an empirical library, empirical spectra will be added to it only if they are high quality
Thread number set to 16
Maximum number of variable modifications set to 5
Modification UniMod:1 with mass delta 42.0106 at *n will be considered as variable
Modification UniMod:35 with mass delta 15.9949 at M will be considered as variable
Modification UniMod:2114 with mass delta 72.0211 at K will be considered as variable
DIA-NN will automatically optimise the mass accuracy for the first run of the experiment, use this mode for preliminary analyses only
WARNING: incorrect settings, the in silico-predicted library must be generated in a separate pipeline step and then used to process the raw data, now without activating FASTA digest
WARNING: peptidoform scoring enabled because variable modifications have been declared; to disable, use --no-peptidoforms
The following variable modifications will be localised: UniMod:1 UniMod:35 UniMod:2114 

2 files will be processed
[0:00] Loading FASTA /public/share//database/2023-02-17-decoys-contam-uniprot_human.fasta.fas
[18:25] Processing FASTA
[18:51] Assembling elution groups
[19:57] 36199194 precursors generated
[19:57] Gene names missing for some isoforms
[19:57] Library contains 80668 proteins, and 25933 genes
[20:34] [22:02] [42:27] [44:30] [45:16] [45:45] Saving the library to report-lib.predicted.speclib
[47:01] Initialising library
[50:18] Loading spectral library report-lib.predicted.speclib
[50:45] Library annotated with sequence database(s): /public/share//database/2023-02-17-decoys-contam-uniprot_human.fasta.fas
[51:02] Spectral library loaded: 149399 protein isoforms, 284130 protein groups and 36199194 precursors in 12201974 elution groups.
[51:02] Loading protein annotations from FASTA /public/share//database/2023-02-17-decoys-contam-uniprot_human.fasta.fas
[51:03] Annotating library proteins with information from the FASTA database
[51:03] Gene names missing for some isoforms
[51:03] Library contains 80668 proteins, and 25933 genes
[51:48] Initialising library

First pass: generating a spectral library from DIA data

[54:50] File #1/2
[54:50] Loading run /public/share//OptDiann/mzML/FWMS20252257_FA_2.mzML
[55:14] Pre-processing...
[55:55] 2223 MS1 and 247711 MS2 scans in 741 (inferred) and 741 (encoded) cycles, 19541024 precursors in range
[56:05] Calibrating with mass accuracies 20 (MS1), 25 (MS2)
[65:24] RT window set to 1.68797
[65:24] Peak width: 3.08
[65:24] Scan window radius set to 6
[65:26] Recommended MS1 mass accuracy setting: 3 ppm
[76:22] Optimised mass accuracy: 8 ppm
[80:29] Searching decoys
[84:23] Main search
[92:00] Removing low confidence identifications
[92:45] Removing interfering precursors
[93:21] Training neural networks on 80884 target and 47082 decoy PSMs
[94:34] Training neural networks on 80884 target and 44941 decoy PSMs
[95:13] Number of IDs at 0.01 FDR: 39933
[95:16] Precursors at 1% peptidoform FDR: 38526
[95:20] Calculating protein q-values
[95:23] Number of genes identified at 1% FDR: 4922 (precursor-level), 4335 (protein-level) (inference performed using proteotypic peptides only)
[95:24] Quantification
[95:26] Precursors with scored PTMs at 1% FDR: 1227 out of 1359 considered
[95:26] Precursors with all scored PTM sites unoccupied at 1% FDR: 37299
[95:26] Precursors with PTMs localised (when required) with > 90% confidence: 903 out of 1227

[95:29] File #2/2
[95:29] Loading run /public/share//OptDiann/mzML/FWMS20252257_FA_1.mzML
[95:35] Pre-processing...
[96:12] 689 MS1 and 76754 MS2 scans in 230 (inferred) and 230 (encoded) cycles, 19541024 precursors in range
[96:22] Calibrating with mass accuracies 21 (MS1), 25 (MS2)
[102:34] RT window set to 0.489203
[102:36] Recommended MS1 mass accuracy setting: 3.4 ppm
[105:53] Searching decoys
[107:37] Main search
[111:14] Removing low confidence identifications
[111:54] Removing interfering precursors
[112:30] Training neural networks on 59052 target and 34713 decoy PSMs
[113:31] Training neural networks on 59052 target and 32899 decoy PSMs
[113:59] Number of IDs at 0.01 FDR: 26341
[114:02] Precursors at 1% peptidoform FDR: 25218
[114:05] Calculating protein q-values
[114:08] Number of genes identified at 1% FDR: 3825 (precursor-level), 3446 (protein-level) (inference performed using proteotypic peptides only)
[114:09] Quantification
[114:10] Precursors with scored PTMs at 1% FDR: 877 out of 962 considered
[114:10] Precursors with all scored PTM sites unoccupied at 1% FDR: 24341
[114:10] Precursors with PTMs localised (when required) with > 90% confidence: 666 out of 877

[114:12] Cross-run analysis
[114:12] Reading quantification information: 2 files
[116:01] Quantifying peptides
[116:29] Assembling protein groups
[116:36] Quantifying proteins
[116:36] Calculating q-values for protein and gene groups
[116:41] Calculating global q-values for protein and gene groups
[116:42] Protein groups with global q-value <= 0.01: 4474
[116:43] Compressed report saved to report-first-pass.parquet. Use R 'arrow' or Python 'PyArrow' package to process
[116:43] Saving precursor levels matrix
[116:43] Precursor levels matrix (1% precursor and protein group FDR) saved to report-first-pass.pr_matrix.tsv.
[116:43] Manifest saved to report-first-pass.manifest.txt
[116:43] Stats report saved to report-first-pass.stats.tsv
[116:45] Generating spectral library:
[116:46] 41120 target and 413 decoy precursors saved
[116:47] Spectral library saved to report-lib.parquet

[117:06] Loading spectral library report-lib.parquet
[117:06] Spectral library loaded: 6296 protein isoforms, 5700 protein groups and 41533 precursors in 36891 elution groups.
[117:06] Loading protein annotations from FASTA /public/share//database/2023-02-17-decoys-contam-uniprot_human.fasta.fas
[117:07] Annotating library proteins with information from the FASTA database
[117:07] Gene names missing for some isoforms
[117:07] Library contains 5769 proteins, and 5594 genes
[117:07] Initialising library
[117:08] Saving the library to report-lib.parquet.skyline.speclib


Second pass: using the newly created spectral library to reanalyse the data

[117:08] File #1/2
[117:08] Loading run /public/share//OptDiann/mzML/FWMS20252257_FA_2.mzML
[117:32] Pre-processing...
[117:35] 2223 MS1 and 247711 MS2 scans in 741 (inferred) and 741 (encoded) cycles, 41120 precursors in range
[117:35] Calibrating with mass accuracies 20 (MS1), 25 (MS2)
[117:36] RT window set to 0.472253
[117:36] Recommended MS1 mass accuracy setting: 4.5 ppm
[117:37] Searching decoys
[117:37] Main search
[117:38] Removing low confidence identifications
[117:40] Removing interfering precursors
[117:41] Training neural networks on 35404 target and 16832 decoy PSMs
[117:54] Training neural networks on 35404 target and 18727 decoy PSMs
[118:07] Number of IDs at 0.01 FDR: 38233
[118:07] Precursors at 1% peptidoform FDR: 37788
[118:07] Calculating protein q-values
[118:07] Number of genes identified at 1% FDR: 4764 (precursor-level), 3768 (protein-level) (inference performed using proteotypic peptides only)
[118:07] Quantification
[118:08] Precursors with scored PTMs at 1% FDR: 1170 out of 1177 considered
[118:08] Precursors with all scored PTM sites unoccupied at 1% FDR: 36618
[118:08] Precursors with PTMs localised (when required) with > 90% confidence: 839 out of 1170

[118:09] File #2/2
[118:09] Loading run /public/share//OptDiann/mzML/FWMS20252257_FA_1.mzML
[118:15] Pre-processing...
[118:16] 689 MS1 and 76754 MS2 scans in 230 (inferred) and 230 (encoded) cycles, 41120 precursors in range
[118:16] Calibrating with mass accuracies 21 (MS1), 24 (MS2)
[118:17] RT window set to 0.207608
[118:17] Recommended MS1 mass accuracy setting: 6 ppm
[118:17] Searching decoys
[118:18] Main search
[118:18] Removing low confidence identifications
[118:20] Removing interfering precursors
[118:21] Training neural networks on 34082 target and 17483 decoy PSMs
[118:34] Training neural networks on 34083 target and 18719 decoy PSMs
[118:46] Number of IDs at 0.01 FDR: 30858
[118:47] Precursors at 1% peptidoform FDR: 29613
[118:47] Calculating protein q-values
[118:47] Number of genes identified at 1% FDR: 3991 (precursor-level), 3388 (protein-level) (inference performed using proteotypic peptides only)
[118:47] Quantification
[118:47] Precursors with scored PTMs at 1% FDR: 1001 out of 1056 considered
[118:47] Precursors with all scored PTM sites unoccupied at 1% FDR: 28612
[118:47] Precursors with PTMs localised (when required) with > 90% confidence: 717 out of 1001

[118:47] Cross-run analysis
[118:47] Reading quantification information: 2 files
[118:48] Quantifying peptides
WARNING: QuantUMS requires 6 or more runs for the optimisation of its hyperparameters to perform best.
[119:13] Quantification parameters: 0.312589, 0.00446964, 0.00200004, 0.304366, 0.0481276, 0.159607, 0.280886, 0.0582567, 0.0776827, 0.110645, 0.0488168, 0.0501528, 0.654398, 0.0502779, 0.0612614, 0.00950628
[119:24] Quantifying proteins
[119:24] Calculating q-values for protein and gene groups
[119:24] Calculating global q-values for protein and gene groups
[119:24] Protein groups with global q-value <= 0.01: 4252
[119:25] Compressed report saved to report.parquet. Use R 'arrow' or Python 'PyArrow' package to process
[119:25] Saving precursor levels matrix
[119:25] Precursor levels matrix (1% precursor and protein group FDR) saved to report.pr_matrix.tsv.
[119:25] Saving protein group levels matrix
[119:25] Protein groups matrix saved to report.pg_matrix.tsv.
[119:25] Saving gene group levels matrix
[119:25] Gene groups matrix saved to report.gg_matrix.tsv.
[119:25] Saving unique genes levels matrix
[119:25] Unique genes matrix saved to report.unique_genes_matrix.tsv.
[119:25] Manifest saved to report.manifest.txt
[119:25] Stats report saved to report.stats.tsv

`
